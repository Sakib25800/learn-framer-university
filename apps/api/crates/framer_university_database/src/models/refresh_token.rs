use chrono::{DateTime, Duration, Utc};
use sqlx::PgPool;
use uuid::Uuid;

use crate::DbResult;

#[derive(Debug, Clone)]
pub struct RefreshTokenModel {
    pub id: Uuid,
    pub user_id: Uuid,
    pub token: String,
    pub expires: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

#[derive(Debug, Clone)]
pub struct RefreshTokens {
    pool: PgPool,
}

impl RefreshTokens {
    pub fn new(pool: PgPool) -> Self {
        Self { pool }
    }

    pub async fn create(&self, user_id: Uuid, expires_in_days: i64) -> DbResult<RefreshTokenModel> {
        let expires = Utc::now() + Duration::days(expires_in_days);

        let token = sqlx::query_as!(
            RefreshTokenModel,
            r#"
            INSERT INTO refresh_tokens (user_id, expires)
            VALUES ($1, $2)
            RETURNING
                id,
                user_id,
                token,
                expires,
                created_at,
                updated_at
            "#,
            user_id,
            expires
        )
        .fetch_one(&self.pool)
        .await?;

        Ok(token)
    }
}
