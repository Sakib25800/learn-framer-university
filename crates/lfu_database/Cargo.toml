[package]
name = "lfu_database"
version = "0.1.0"
edition = "2024"

[lints]
workspace = true

[dependencies]
sqlx = { version = "0.8.5", features = ["runtime-tokio", "tls-rustls", "postgres", "chrono", "uuid"] }
uuid = { version = "1.16.0", features = ["v4"]}
chrono = "0.4.40"
