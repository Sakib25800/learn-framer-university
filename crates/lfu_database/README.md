# lfu_database

This package contains the database schema as derived by `diesel print-schema` from the databse after all migrations have been applied.

After creating new migrations (via `diesel migration generate`), update the schema by executing:
```console
diesel print-schema > crates/lfu_database/src/schema.rs
```
