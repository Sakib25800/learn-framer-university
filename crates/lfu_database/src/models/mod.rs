pub mod account;
pub mod refresh_token;
pub mod user;
pub mod verification_token;
