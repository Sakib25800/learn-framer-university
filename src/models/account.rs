use crate::schema::accounts;
use chrono::NaiveDateTime;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use diesel_derive_enum::DbEnum;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Queryable, Identifiable, Selectable, AsChangeset)]
#[diesel(table_name = accounts)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct Account {
    pub id: i64,
    pub user_id: i64,
    pub auth_method: AuthMethod,
    pub provider: String,
    pub provider_account_id: String,
    pub refresh_token: Option<String>,
    pub access_token: Option<String>,
    pub expires_at: Option<i64>,
    pub token_type: Option<String>,
    pub scope: Option<String>,
    pub id_token: Option<String>,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Debug, DbEnum, Serialize, Deserialize)]
#[ExistingTypePath = "crate::schema::sql_types::AuthMethodEnum"]
pub enum AuthMethod {
    #[db_rename = "oauth"]
    OAuth,
    #[db_rename = "email"]
    Email,
    #[db_rename = "credentials"]
    Credentials,
    #[db_rename = "oidc"]
    OIDC,
    #[db_rename = "sms"]
    SMS,
}

impl Default for AuthMethod {
    fn default() -> Self {
        AuthMethod::Email
    }
}

impl Account {
    pub async fn find(id: i64, conn: &mut AsyncPgConnection) -> QueryResult<Account> {
        accounts::table.find(id).first(conn).await
    }
}

#[derive(Insertable, Debug)]
#[diesel(table_name = accounts)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct NewAccount<'a> {
    pub user_id: i64,
    pub auth_method: AuthMethod,
    pub provider: &'a str,
    pub provider_account_id: &'a str,
    pub refresh_token: Option<&'a str>,
    pub access_token: Option<&'a str>,
    pub expires_at: Option<i64>,
    pub token_type: Option<&'a str>,
    pub scope: Option<&'a str>,
    pub id_token: Option<&'a str>,
}

impl<'a> NewAccount<'a> {
    pub fn new(
        user_id: i64,
        auth_method: AuthMethod,
        provider: &'a str,
        provider_account_id: &'a str,
        refresh_token: Option<&'a str>,
        access_token: Option<&'a str>,
        expires_at: Option<i64>,
        token_type: Option<&'a str>,
        scope: Option<&'a str>,
        id_token: Option<&'a str>,
    ) -> Self {
        NewAccount {
            user_id,
            auth_method,
            provider,
            provider_account_id,
            refresh_token,
            access_token,
            expires_at,
            token_type,
            scope,
            id_token,
        }
    }

    pub async fn create(&self, conn: &mut AsyncPgConnection) -> QueryResult<Account> {
        diesel::insert_into(accounts::table)
            .values(self)
            .returning(Account::as_returning())
            .get_result(conn)
            .await
    }
}
