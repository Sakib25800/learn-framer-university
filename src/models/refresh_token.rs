use crate::schema::refresh_tokens;
use chrono::{NaiveDateTime, Utc};
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use serde::{Deserialize, Serialize};

#[derive(
    Debug, Clone, Serialize, Deserialize, Queryable, Identifiable, Selectable, AsChangeset,
)]
#[diesel(table_name = refresh_tokens)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct RefreshToken {
    pub id: i64,
    pub user_id: i64,
    pub token: String,
    pub expires: NaiveDateTime,
    pub created_at: NaiveDateTime,
}

impl RefreshToken {
    pub async fn find(id: i64, conn: &mut AsyncPgConnection) -> QueryResult<RefreshToken> {
        refresh_tokens::table.find(id).first(conn).await
    }
}

#[derive(Insertable, Debug)]
#[diesel(table_name = refresh_tokens)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct NewRefreshToken {
    pub user_id: i64,
    pub token: String,
    pub expires: NaiveDateTime,
}

impl NewRefreshToken {
    pub fn new(user_id: i64, expires_in_days: i64) -> Self {
        let token = uuid::Uuid::new_v4().to_string();
        let expires = (Utc::now() + chrono::Duration::days(expires_in_days)).naive_utc();

        Self {
            user_id,
            token,
            expires,
        }
    }

    pub async fn create(&self, conn: &mut AsyncPgConnection) -> QueryResult<RefreshToken> {
        diesel::insert_into(refresh_tokens::table)
            .values(self)
            .returning(RefreshToken::as_returning())
            .get_result(conn)
            .await
    }
}
