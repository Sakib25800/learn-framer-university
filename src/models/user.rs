use crate::schema::users;
use chrono::NaiveDateTime;
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use serde::{Deserialize, Serialize};

#[derive(
    Debug, Clone, Serialize, Deserialize, Queryable, Identifiable, Selectable, AsChangeset,
)]
#[diesel(table_name = users)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct User {
    pub id: i64,
    pub email: String,
    pub email_verified: Option<NaiveDateTime>,
    pub image: Option<String>,
    pub is_admin: bool,
    pub last_active_at: NaiveDateTime,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

impl User {
    pub async fn find(id: i64, conn: &mut AsyncPgConnection) -> QueryResult<User> {
        users::table.find(id).first(conn).await
    }

    pub async fn find_by_email(email_str: &str, conn: &mut AsyncPgConnection) -> QueryResult<User> {
        users::table
            .filter(users::email.eq(email_str))
            .first(conn)
            .await
    }

    pub async fn verify_email(id: i64, conn: &mut AsyncPgConnection) -> QueryResult<User> {
        diesel::update(users::table)
            .filter(users::id.eq(id))
            .set(users::email_verified.eq(Some(chrono::Utc::now().naive_utc())))
            .returning(User::as_returning())
            .get_result(conn)
            .await
    }

    pub async fn update_email_and_verify(
        id: i64,
        new_email: &str,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<User> {
        diesel::update(users::table)
            .filter(users::id.eq(id))
            .set((
                users::email.eq(new_email),
                users::email_verified.eq(Some(chrono::Utc::now().naive_utc())),
            ))
            .returning(User::as_returning())
            .get_result(conn)
            .await
    }
}

#[derive(Insertable, Debug)]
#[diesel(table_name = users)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct NewUser<'a> {
    pub email: &'a str,
    pub email_verified: Option<NaiveDateTime>,
}

impl<'a> NewUser<'a> {
    pub fn new(email: &'a str) -> Self {
        NewUser {
            email,
            email_verified: None,
        }
    }

    pub async fn create(&self, conn: &mut AsyncPgConnection) -> QueryResult<User> {
        diesel::insert_into(users::table)
            .values(self)
            .returning(User::as_returning())
            .get_result(conn)
            .await
    }
}
