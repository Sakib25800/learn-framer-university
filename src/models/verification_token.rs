use crate::schema::verification_tokens;
use chrono::{DateTime, Utc};
use diesel::prelude::*;
use diesel_async::{AsyncPgConnection, RunQueryDsl};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Queryable, Selectable, AsChangeset)]
#[diesel(table_name = verification_tokens)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct VerificationToken {
    pub identifier: String,
    pub token: String,
    pub expires: DateTime<Utc>,
    pub created_at: DateTime<Utc>,
}

impl VerificationToken {
    pub async fn find_by_token(
        token: &str,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<VerificationToken> {
        verification_tokens::table
            .filter(verification_tokens::token.eq(token))
            .first(conn)
            .await
    }

    pub async fn delete(
        identifier: &str,
        token: &str,
        conn: &mut AsyncPgConnection,
    ) -> QueryResult<usize> {
        diesel::delete(verification_tokens::table)
            .filter(verification_tokens::identifier.eq(identifier))
            .filter(verification_tokens::token.eq(token))
            .execute(conn)
            .await
    }
}

#[derive(Insertable, Debug)]
#[diesel(table_name = verification_tokens)]
#[diesel(check_for_backend(diesel::pg::Pg))]
pub struct NewVerificationToken {
    pub identifier: String,
    pub token: String,
    pub expires: chrono::NaiveDateTime,
}

impl NewVerificationToken {
    pub fn new(identifier: String, expires_in_hours: i64) -> Self {
        let token = uuid::Uuid::new_v4().to_string();
        let expires = Utc::now().naive_utc() + chrono::Duration::hours(expires_in_hours);

        Self {
            identifier,
            token,
            expires,
        }
    }

    pub async fn create(&self, conn: &mut AsyncPgConnection) -> QueryResult<VerificationToken> {
        diesel::insert_into(verification_tokens::table)
            .values(self)
            .returning(VerificationToken::as_returning())
            .get_result(conn)
            .await
    }
}
