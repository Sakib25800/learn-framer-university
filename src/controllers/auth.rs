use axum::Json;
use chrono::Utc;
use serde::Deserialize;
use utoipa::ToSchema;
use validator::Validate;

use lfu_database::models::{
    refresh_token::NewRefreshToken,
    user::{NewUser, User},
    verification_token::{NewVerificationToken, VerificationToken},
};

use crate::{
    app::AppState,
    auth::generate_access_token,
    middleware::{json::JsonBody, path::ValidatedPath},
    util::errors::{unauthorized, AppErrorResponse, AppResult},
    views::{MessageResponse, VerifiedEmailResponse},
};

#[derive(Deserialize, Validate, ToSchema)]
pub struct AuthSignInBody {
    #[validate(email)]
    email: String,
}

#[derive(Deserialize, Validate)]
pub struct VerifyEmailQueryParams {
    #[validate(length(min = 1))]
    token: String,
}

/// Sign in the user and send a sign-in email
#[utoipa::path(
    post,
    path = "/v1/auth/signin",
    tag = "auth",
    request_body = AuthSignInBody,
    responses(
        (status = 200, body = MessageResponse, description = "successful operation"),
        (status = 400, body = AppErrorResponse, description = "failed operation"),
    ),
)]
pub async fn signin(
    state: AppState,
    JsonBody(body): JsonBody<AuthSignInBody>,
) -> AppResult<Json<MessageResponse>> {
    let mut conn = state.database.get().await?;
    let email = &body.email;

    match User::find_by_email(email, &mut conn).await {
        Ok(user) => {
            // Revoke all previously generated verification tokens
            VerificationToken::delete_all(&user.email, &mut conn).await?;
            user
        }
        Err(_) => NewUser::new(&body.email, false).insert(&mut conn).await?,
    };

    let verification_token = NewVerificationToken::new(
        email.to_owned(),
        state.config.email_verification_expiration_hours,
    )
    .create(&mut conn)
    .await?;

    let signin_email = AuthSignInEmail {
        app_url: &state.config.app_url,
        token: &verification_token.token,
    };

    state.emails.send(&body.email, signin_email).await?;

    Ok(Json(MessageResponse {
        message: "We've sent you an email".to_owned(),
    }))
}

#[derive(Deserialize, Validate)]
pub struct AuthSignInParams {
    #[validate(length(min = 1))]
    pub token: String,
}

/// Verify the user's email
#[utoipa::path(
    get,
    path = "/v1/auth/continue/{token}",
    params(
        ("token" = String, Path, description = "Token used to verify email")
    ),
    responses(
        (status = 200, body = VerifiedEmailResponse, description = "successful operation"),
        (status = 400, body = AppErrorResponse, description = "failed operation"),
    ),
    tag = "auth",
)]
pub async fn continue_signin(
    state: AppState,
    ValidatedPath(params): ValidatedPath<AuthSignInParams>,
) -> AppResult<Json<VerifiedEmailResponse>> {
    let token = params.token;
    let mut conn = state.database.get().await?;

    let verification_token = match VerificationToken::find_by_token(&token, &mut conn).await {
        Ok(token) => token,
        Err(_) => return Err(unauthorized("Invalid verification token")),
    };

    if verification_token.expires < Utc::now() {
        return Err(unauthorized("Verification token has expired"));
    }

    let user = match User::find_by_email(&verification_token.identifier, &mut conn).await {
        Ok(user) => user,
        Err(_) => return Err(unauthorized("Email does not exist")),
    };

    let crate::config::Server {
        jwt_secret,
        jwt_access_token_expiration_hours,
        jwt_refresh_token_expiration_days,
        ..
    } = state.config.as_ref();
    let access_token = generate_access_token(jwt_secret, jwt_access_token_expiration_hours, &user)?;
    let refresh_token = NewRefreshToken::new(user.id, *jwt_refresh_token_expiration_days)
        .insert(&mut conn)
        .await?;

    // Set user email as verified
    User::verify_email(user.id, &mut conn).await?;

    // Delete the used verification token
    VerificationToken::delete(&verification_token.identifier, token.as_str(), &mut conn).await?;

    Ok(Json(VerifiedEmailResponse {
        access_token,
        refresh_token: refresh_token.token,
    }))
}

pub struct AuthSignInEmail<'a> {
    pub app_url: &'a str,
    pub token: &'a str,
}

impl crate::email::Email for AuthSignInEmail<'_> {
    fn subject(&self) -> String {
        "Framer University: Please confirm your email address".into()
    }

    fn body(&self) -> String {
        format!(
            "Hey there! Welcome to Framer University.\nPlease click the link below to sign in: {app_url}/api/continue/{token}",
            app_url = self.app_url,
            token = self.token,
        )
    }
}
