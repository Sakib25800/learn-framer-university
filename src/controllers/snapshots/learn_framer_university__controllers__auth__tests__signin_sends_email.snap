---
source: src/controllers/auth.rs
expression: app.emails_snapshot().await
---
To: foo@example.com
From: learn.framer.university <noreply@learn.framer.university>
Subject: Activation link for Framer University
Content-Type: text/plain; charset=utf-8
Content-Transfer-Encoding: quoted-printable

Hey there! Welcome to Framer University.
Please click the link below to sign in: https://learn.framer.university/api/continue/[token]
